bitcoin-core-sv2/src/lib.rs

```diff
@@ -0,0 +1,1138 @@
+
+        let template_block_bytes = template_block_request
+            .send()
+            .promise
+            .await?
+            .get()?
+            .get_result()?
+            .to_vec();
+
+        // Deserialize the complete block template from Bitcoin Core's serialization format
+        tracing::debug!(
+            "Deserializing block template ({} bytes)",
+            template_block_bytes.len()
+        );
+        let block: Block = deserialize(&template_block_bytes)?;
+        tracing::debug!(
+            "Block deserialized - prev_hash from header: {:?}",
+            block.header.prev_blockhash
+        );
+
+        // Create the template data structure
+        let template_data = TemplateData::new(template_id, block, template_ipc_client);
+        tracing::debug!("TemplateData created successfully");
+
+        // todo: remove this once https://github.com/bitcoin/bitcoin/issues/33575 is implemented
+        self.pending_fetch.store(false, Ordering::SeqCst);
+
+        Ok(template_data)
+    }
+
+    async fn new_template_ipc_client(
+        &self,
+        coinbase_output_max_additional_size: u32,
+        coinbase_output_max_additional_sigops: u16,
+    ) -> Result<BlockTemplateIpcClient, BitcoinCoreSv2Error> {
+        tracing::debug!(
+            "new_template_ipc_client() called - max_size: {}, max_sigops: {}",
+            coinbase_output_max_additional_size,
+            coinbase_output_max_additional_sigops
+        );
+
+        let mut template_ipc_client_request = self.mining_ipc_client.create_new_block_request();
+        let mut template_ipc_client_request_options =
+            match template_ipc_client_request.get().get_options() {
+                Ok(options) => options,
+                Err(e) => {
+                    tracing::error!("Failed to get template IPC client request options: {}", e);
+                    return Err(BitcoinCoreSv2Error::CapnpError(e));
+                }
+            };
+
+        let coinbase_weight = (coinbase_output_max_additional_size * 4) as u64;
+        let block_reserved_weight = coinbase_weight.max(2000); // 2000 is the minimum block reserved weight
+        tracing::debug!("Setting block_reserved_weight: {}", block_reserved_weight);
+        template_ipc_client_request_options.set_block_reserved_weight(block_reserved_weight);
+        template_ipc_client_request_options.set_coinbase_output_max_additional_sigops(
+            coinbase_output_max_additional_sigops as u64,
+        );
+        template_ipc_client_request_options.set_use_mempool(true);
+
+        tracing::debug!("Sending createNewBlock request to Bitcoin Core");
+        let template_ipc_client_response = match template_ipc_client_request.send().promise.await {
+            Ok(response) => response,
+            Err(e) => {
+                tracing::error!("Failed to send template IPC client request: {}", e);
+                return Err(BitcoinCoreSv2Error::CapnpError(e));
+            }
+        };
```

Comment:
We can implement Display for our error types so that this message is automatically included when the error is formatted.
```suggestion
        let template_ipc_client_response = template_ipc_client_request.send().promise.await?; 
```